### ISSUE TITLE
Sessions are disabled for this entry point

### ISSUE THREAD

**Issue by Nikerabbit** [SemanticMediaWiki/SemanticMediaWiki#1867]
### Setup and configuration

https://dev.translatewiki.net/wiki/Special:Version
- SMW version: 2.4.1
- MW version: 1.28-master
- DB (MySQL etc.):
### Issue

Produces a [stack trace](https://www.semantic-mediawiki.org/wiki/Help:Identifying_bugs) and outputs:

```
2016-09-21 08:42:47 translatewiki.net dev_translatewiki_net-bw_: [79dcb8da769fe5d1716782cb] /w/load.php?debug=false&lang=fi&modules=mobile.search%7Cskins.minerva.notifications%2Cscripts%2Ctoggling&skin=minerva&version=1ytnjkt   BadMethodCallException from line 845 of /www/dev.translatewiki.net/docroot/w/includes/session/SessionManager.php: Sessions are disabled for this entry point
#0 /www/dev.translatewiki.net/docroot/w/includes/session/SessionManager.php(194): MediaWiki\Session\SessionManager->getSessionFromInfo(MediaWiki\Session\SessionInfo, WebRequest)
#1 /www/dev.translatewiki.net/docroot/w/includes/WebRequest.php(734): MediaWiki\Session\SessionManager->getSessionForRequest(WebRequest)
#2 /www/dev.translatewiki.net/docroot/w/includes/user/User.php(1198): WebRequest->getSession()
#3 /www/dev.translatewiki.net/docroot/w/includes/user/User.php(403): User->loadFromSession()
#4 /www/dev.translatewiki.net/docroot/w/includes/user/User.php(5120): User->load()
#5 /www/dev.translatewiki.net/docroot/w/includes/user/User.php(2763): User->loadOptions()
#6 /www/dev.translatewiki.net/docroot/w/includes/context/RequestContext.php(363): User->getOption(string)
#7 /www/dev.translatewiki.net/docroot/w/includes/StubObject.php(207): RequestContext->getLanguage()
#8 /www/dev.translatewiki.net/docroot/w/includes/StubObject.php(169): StubUserLang->_newObject()
#9 /www/dev.translatewiki.net/docroot/w/includes/StubObject.php(112): StubObject->_unstub(string, integer)
#10 /www/dev.translatewiki.net/docroot/w/includes/StubObject.php(139): StubObject->_call(string, array)
#11 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/DataTypeRegistry.php(127): StubObject->__call(string, array)
#12 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/DataValueFactory.php(58): SMW\DataTypeRegistry::getInstance()
#13 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/InTextAnnotationParser.php(88): SMW\DataValueFactory::getInstance()
#14 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/ApplicationFactory.php(235): SMW\InTextAnnotationParser->__construct(SMW\ParserData, SMW\MediaWiki\MagicWordsFinder, SMW\MediaWiki\RedirectTargetFinder)
#15 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/MediaWiki/Hooks/InternalParseBeforeLinks.php(108): SMW\ApplicationFactory->newInTextAnnotationParser(SMW\ParserData)
#16 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/MediaWiki/Hooks/InternalParseBeforeLinks.php(62): SMW\MediaWiki\Hooks\InternalParseBeforeLinks->performUpdate()
#17 /www/dev.translatewiki.net/docroot/w/extensions/SemanticMediaWiki/src/MediaWiki/Hooks/HookRegistry.php(204): SMW\MediaWiki\Hooks\InternalParseBeforeLinks->process()
#18 /www/dev.translatewiki.net/docroot/w/includes/Hooks.php(195): Closure$SMW\MediaWiki\Hooks\HookRegistry::addCallbackHandlers#6(Parser, string, StripState)
#19 /www/dev.translatewiki.net/docroot/w/includes/parser/Parser.php(1262): Hooks::run(string, array)
#20 /www/dev.translatewiki.net/docroot/w/includes/parser/Parser.php(440): Parser->internalParse(string)
#21 /www/dev.translatewiki.net/docroot/w/includes/cache/MessageCache.php(1103): Parser->parse(string, Title, ParserOptions, boolean)
#22 /www/dev.translatewiki.net/docroot/w/includes/Message.php(1146): MessageCache->parse(string, Title, boolean, boolean, LanguageFi)
#23 /www/dev.translatewiki.net/docroot/w/includes/Message.php(828): Message->parseText(string)
#24 /www/dev.translatewiki.net/docroot/w/includes/Message.php(884): Message->toString()
#25 /www/dev.translatewiki.net/docroot/w/extensions/MobileFrontend/includes/modules/MFResourceLoaderParsedMessageModule.php(47): Message->parse()
#26 /www/dev.translatewiki.net/docroot/w/extensions/MobileFrontend/includes/modules/MFResourceLoaderParsedMessageModule.php(78): MFResourceLoaderParsedMessageModule->addParsedMessages(ResourceLoaderContext)
#27 /www/dev.translatewiki.net/docroot/w/includes/resourceloader/ResourceLoaderModule.php(626): MFResourceLoaderParsedMessageModule->getScript(ResourceLoaderContext)
#28 /www/dev.translatewiki.net/docroot/w/includes/resourceloader/ResourceLoaderModule.php(594): ResourceLoaderModule->buildContent(ResourceLoaderContext)
#29 /www/dev.translatewiki.net/docroot/w/includes/resourceloader/ResourceLoader.php(1011): ResourceLoaderModule->getModuleContent(ResourceLoaderContext)
#30 /www/dev.translatewiki.net/docroot/w/includes/resourceloader/ResourceLoader.php(736): ResourceLoader->makeModuleResponse(ResourceLoaderContext, array, array)
#31 /www/dev.translatewiki.net/docroot/w/load.php(46): ResourceLoader->respond(ResourceLoaderContext)
#32 {main}
```

Steps to reproduce the observation (recommendation is to use the [sandbox](http://sandbox.semantic-mediawiki.org)):

Happens with VisualEditor and MobileFrontend. For example when logged in to dev.translatewiki.net (as of now anyway).

I am being the messenger here. From https://phabricator.wikimedia.org/T134562

> ! In T134562#2656041, @Anomie wrote:
> MobileFrontend in that stack trace is doing what it's supposed to do: it's loading the message via the ResourceLoaderContext, which sets the correct language.
> 
> SMW is breaking things in that trace by attempting to access the session user's language during the parsing of the wikitext of a message in a context (load.php) where there is no session and therefore no session user. Chances are it should be using the parser instance's target language instead of the session user's language.

**Comment by JeroenDeDauw** [comment]
Thanks for reporting this issue!

**Comment by mwjames** [comment]
I'd expected someone else to investigate but since it didn't happen allow me to add the following:

> SMW is breaking things in that trace by attempting

This is an ought comment to encourage corporation, especially given that in previous releases there were no such issues and I'm leaning towards the introduction of the `SessionManager` to be the main reason for said issue ( #1867 comes to mind as well).

Anyway, I have taken the time to dissect the issue on the given information:

- SMW should never act on a `Message::parse` (since we are not expecting any valid annotation from that isolated parse process) and the only accessible point to make that decision has been `ParserOptions::getInterfaceMessage`.
- Normally, when the `InternalParseBeforeLinks` hook is triggered SMW checks whether to proceed or not by looking at `ParserOptions::getInterfaceMessage` [0, 1] which is expected to be set when `Message::parse` is used
- If above is an invalid assumption and there are other ways to identify whether a parse was initiated by `Message::parse` (assuming `ParserOptions::getInterfaceMessage` isn't set) then I'd like to hear about it. I'm pretty sure other standard MW processes set the `getInterfaceMessage` correctly and therefore I'm not sure why this doesn't work in connection with `MobileFrontend` [2].
- Since the `InternalParseBeforeLinks` hook is executed (which it shouldn't given above remarks) and in anticipation of possible implications for future changes infused by developers, I provisionally removed the call to the user language from `DataTypeRegistry.php(127): StubObject->__call(string, array)` in [3]   which will only work with changes forthcoming in #1879 and therefore are unlikely to be backported to 2.4

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/InternalParseBeforeLinks.php#L73
[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/src/MediaWiki/Hooks/ParserAfterTidy.php#L60
[2] https://github.com/wikimedia/mediawiki-extensions-MobileFrontend/blob/master/includes/modules/MFResourceLoaderParsedMessageModule.php#L47
[3] https://github.com/SemanticMediaWiki/SemanticMediaWiki/pull/1879/files#diff-1c181fb96c42814c7c3a02fe5770c569R125

**Comment by mwjames** [comment]
As I was unable to attain additional answers on raised questions above, I'm inclined to close this for the reasons already stated.

**Comment by Nikerabbit** [comment]
I confirm that 219a45025fed271e2ffa3aba0ade3c3f29eec7ec, at least on quick testing, makes this error go away.

**Comment by Nikerabbit** [comment]
There are some comments to your questions at https://phabricator.wikimedia.org/T134562#2814683

**Comment by kghbln** [comment]
@Nikerabbit Thanks for following up on this as well as providing the pointer!

**Comment by waandershz** [comment]
I still encounter this error: https://www.mediawiki.org/wiki/Topic:Trnjlmc7a4oobq1e

**Comment by mwjames** [comment]
> I still encounter this error: https://www.mediawiki.org/wiki/Topic:Trnjlmc7a4oobq1e

For the future, please create a new issue. I know that this might be confusing but keeping issues distinct helps us as well as others to concentrate on problems at hand.

As for the linked topic, you have described this with "It's working fine till we use the Visual Editor-extension. Console of browser gives this error ... BadMethodCallException from line 845 of /home/hz01/mediawiki/core/includes/session/SessionManager.php: Sessions are disabled for this entry point"

On further elaboration you stated that "This message disappears as soon as I remove "require_once "$IP/extensions/VisualEditor/VisualEditor.php"" from LocalSettings.php.".

How does "VisualEditor" and SMW are linked together in this particular case?

Of course, I can see from the description one may let to believe this to be a unique issue but doing a quick search on phabricator.wikimedia.org reveals https://phabricator.wikimedia.org/T146686 ("VE fails startup module with "Sessions are disabled for this entry point").

PS: When presenting or describing an issue, please ensure to provide a full stack trace and consulting the  phabricator.wikimedia.org search is recommended as well.

**Comment by waandershz** [comment]
Thanks. I will keep it in mind.

I don fully understand the second part of your reply. I did [post](https://www.mediawiki.org/wiki/Topic:Trnjlmc7a4oobq1e) the issue at the Visual Editor discussion page but  the answer was it's a SMW issue, which should be fixed, but though I use SMW 2.5.2 we encounter this error.

How do I provide a full stack trace using Chrome?

Thanks for your help!

**Comment by mwjames** [comment]
Hi,

> I don fully understand the second part of your reply. I did post the issue at the Visual Editor discussion page but the answer was it's a SMW issue, which should be fixed, but though I

I linked the issue that is related to your problem in regards to VE.
As I said previously, the issue is unrelated to SMW, please review the
linked phab ticket and the recommendations posted their.

> How do I provide a full stack trace using Chrome?

https://www.semantic-mediawiki.org/wiki/Help:Identifying_bugs

Cheers
